### ISSUE TITLE
Bug in POSIX::strerror on 5.21.7

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14370]
Migrated from [rt.perl.org#123503](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123503) (status was 'resolved')

Searchable as RT123503$

**Comment by p5pRT** [comment]
### From @peczenyj

Hello\,

I find one bug on POSIX&#8203;::strerror on 5\.21\.7 \- it erase the $\! variable\!

I discover this based on this ticket&#8203;:

https://github.com/Weborama/Riak-Light/issues/18

I try to reproduce the error using the 5\.21\.7 and I discover one strange behavior on POSIX&#8203;::strerror

IF I do this

\< perl code >
use POSIX qw\(ETIMEDOUT strerror\);
$\! = ETIMEDOUT;
my $errmsg = strerror\(ETIMEDOUT\);
\< / perl code >

after the strerror\, the content of $\! was totally erased\. and it has a bad impact on my tests\.

Seems the origin of this problem is this change between 5\.21\.6 and 5\.21\.7

\- strerror  => 'errno => local $\! = $\_\[0\]; "$\!"'\,
\+ strerror  => 'errno => use locale; local $\! = $\_\[0\]; "$\!"'\,

I attach two files&#8203;:

one is a propose to test this bug\. it works for 5\.12\.2 and fails for 5\.21\.7

another is a more complete test\, to show what I am doing\.

**Comment by p5pRT** [comment]
### From @peczenyj

use strict;
use warnings;
use Test&#8203;::More;
use POSIX qw\(ETIMEDOUT strerror\);

$\! = ETIMEDOUT;
my $errmsg = strerror\(ETIMEDOUT\);

ok $errmsg\, "should convert the ETIMEDOUT to text";
ok $\!\, "should not erase the value of ERRNO";

done\_testing;

**Comment by p5pRT** [comment]
### From @peczenyj

package Subject;
use Data&#8203;::Dumper;

sub new \{
&nbsp;  bless \{\};
\}

sub ping \{
&nbsp;  die "lol \($\!\)"
\}

package main;

use strict;
use warnings;
use Data&#8203;::Dumper;
use Test&#8203;::More;
use Test&#8203;::Exception;
use POSIX qw\(ETIMEDOUT strerror\);


&nbsp;        my $client = Subject\->new\(\);

&nbsp;        $\! = ETIMEDOUT;
&nbsp;        \#debug
&nbsp;        diag Dumper \{ BEFORE => \{ error => $\! \}\} ;
&nbsp;        my $errmsg = strerror\(ETIMEDOUT\);
&nbsp;        diag Dumper \{ AFTER => \{ error => $\! \}\} ;
&nbsp;        throws\_ok \{ $client\->ping\(\) \} qr/lol \\\($errmsg\\\)/\,
&nbsp;          "should die";

done\_testing;

**Comment by p5pRT** [comment]
### From @eserte

Dana Pet 26\. Pro 2014\, 05&#8203;:29&#8203;:25\, tiago\.peczenyj@&#8203;gmail\.com reče&#8203;:
> Hello\,
> 
> I find one bug on POSIX&#8203;::strerror on 5\.21\.7 \- it erase the $\!
> variable\!
> 
> I discover this based on this ticket&#8203;:
> 
> https://github.com/Weborama/Riak-Light/issues/18
> 
> I try to reproduce the error using the 5\.21\.7 and I discover one
> strange behavior on POSIX&#8203;::strerror
> 
> IF I do this
> 
> \< perl code >
> use POSIX qw\(ETIMEDOUT strerror\);
> $\! = ETIMEDOUT;
> my $errmsg = strerror\(ETIMEDOUT\);
> \< / perl code >
> 
> after the strerror\, the content of $\! was totally erased\. and it has a
> bad impact on my tests\.
> 
> Seems the origin of this problem is this change between 5\.21\.6 and
> 5\.21\.7
> 
> \- strerror  => 'errno => local $\! = $\_\[0\]; "$\!"'\,
> \+ strerror  => 'errno => use locale; local $\! = $\_\[0\]; "$\!"'\,
> 
> I attach two files&#8203;:
> 
> one is a propose to test this bug\. it works for 5\.12\.2 and fails for
> 5\.21\.7
> 
> another is a more complete test\, to show what I am doing\.

A possible fix is attached\.

Regards\,
&nbsp;    Slaven

**Comment by p5pRT** [comment]
### From @eserte

<details><summary>0001-properly-localize-in-POSIX-strerror.patch</summary>

```diff
From fea723ade9415b5ae1505bfb981720157de2b365 Mon Sep 17 00:00:00 2001
From: Slaven Rezic <slaven@rezic.de>
Date: Fri, 26 Dec 2014 15:19:22 +0100
Subject: [PATCH] properly localize $! in POSIX::strerror

Just calling 'use locale' may unset $!. This should fix
https://rt.perl.org/Ticket/Display.html?id=123503
---
 ext/POSIX/lib/POSIX.pm | 2 +-
 ext/POSIX/t/wrappers.t | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/ext/POSIX/lib/POSIX.pm b/ext/POSIX/lib/POSIX.pm
index 5695990..f3651e1 100644
--- a/ext/POSIX/lib/POSIX.pm
+++ b/ext/POSIX/lib/POSIX.pm
@@ -151,7 +151,7 @@ my %reimpl = (
     exit      => 'status => CORE::exit($_[0])',
     getenv    => 'name => $ENV{$_[0]}',
     system    => 'command => CORE::system($_[0])',
-    strerror  => 'errno => use locale; local $! = $_[0]; "$!"',
+    strerror  => 'errno => local $!; require locale; locale->import; $! = $_[0]; "$!"',
     strstr    => 'big, little => CORE::index($_[0], $_[1])',
     chmod     => 'mode, filename => CORE::chmod($_[0], $_[1])',
     fstat     => 'fd => CORE::open my $dup, "<&", $_[0]; CORE::stat($dup)', # Gross.
diff --git a/ext/POSIX/t/wrappers.t b/ext/POSIX/t/wrappers.t
index 5b9f6d6..5b38332 100644
--- a/ext/POSIX/t/wrappers.t
+++ b/ext/POSIX/t/wrappers.t
@@ -118,6 +118,7 @@ is_deeply([POSIX::stat($temp_file)], [stat $temp_file], 'stat');
     local $! = 2;
     my $error = "$!";
     is(POSIX::strerror(2), $error, 'strerror');
+    is($!+0, 2, 'previous $! value got restored');
 }
 
 is(POSIX::strstr('BBFRPRAFPGHPP', 'FP'), 7, 'strstr');
-- 
2.1.2


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 26 06&#8203;:21&#8203;:08 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
> Dana Pet 26\. Pro 2014\, 05&#8203;:29&#8203;:25\, tiago\.peczenyj@&#8203;gmail\.com reče&#8203;:
> > Hello\,
> > 
> > I find one bug on POSIX&#8203;::strerror on 5\.21\.7 \- it erase the $\!
> > variable\!
> > 
> > I discover this based on this ticket&#8203;:
> > 
> > https://github.com/Weborama/Riak-Light/issues/18
> > 
> > I try to reproduce the error using the 5\.21\.7 and I discover one
> > strange behavior on POSIX&#8203;::strerror
> > 
> > IF I do this
> > 
> > \< perl code >
> > use POSIX qw\(ETIMEDOUT strerror\);
> > $\! = ETIMEDOUT;
> > my $errmsg = strerror\(ETIMEDOUT\);
> > \< / perl code >
> > 
> > after the strerror\, the content of $\! was totally erased\. and it has a
> > bad impact on my tests\.
> > 
> > Seems the origin of this problem is this change between 5\.21\.6 and
> > 5\.21\.7
> > 
> > \- strerror  => 'errno => local $\! = $\_\[0\]; "$\!"'\,
> > \+ strerror  => 'errno => use locale; local $\! = $\_\[0\]; "$\!"'\,
> > 
> > I attach two files&#8203;:
> > 
> > one is a propose to test this bug\. it works for 5\.12\.2 and fails for
> > 5\.21\.7
> > 
> > another is a more complete test\, to show what I am doing\.
> 
> A possible fix is attached\.

I think this will cause bug \#123171 to recur\.  locale\->import affects the currently\-compiling code\, so it needs to go in a BEGIN block&#8203;: BEGIN \{ local $\!; require locale; locale\->import\} local $\! = $\_\[0\]; "$\!"\.  I haven’t tested any of this\.  Could you test your one\-liner from \#123171 and see whether strerror’s German output is affected by your patch?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @eserte

Dana Pet 26\. Pro 2014\, 07&#8203;:10&#8203;:24\, sprout reče&#8203;:
> On Fri Dec 26 06&#8203;:21&#8203;:08 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
> > Dana Pet 26\. Pro 2014\, 05&#8203;:29&#8203;:25\, tiago\.peczenyj@&#8203;gmail\.com reče&#8203;:
> > > Hello\,
> > >
> > > I find one bug on POSIX&#8203;::strerror on 5\.21\.7 \- it erase the $\!
> > > variable\!
> > >
> > > I discover this based on this ticket&#8203;:
> > >
> > > https://github.com/Weborama/Riak-Light/issues/18
> > >
> > > I try to reproduce the error using the 5\.21\.7 and I discover one
> > > strange behavior on POSIX&#8203;::strerror
> > >
> > > IF I do this
> > >
> > > \< perl code >
> > > use POSIX qw\(ETIMEDOUT strerror\);
> > > $\! = ETIMEDOUT;
> > > my $errmsg = strerror\(ETIMEDOUT\);
> > > \< / perl code >
> > >
> > > after the strerror\, the content of $\! was totally erased\. and it
> > > has a
> > > bad impact on my tests\.
> > >
> > > Seems the origin of this problem is this change between 5\.21\.6 and
> > > 5\.21\.7
> > >
> > > \- strerror  => 'errno => local $\! = $\_\[0\]; "$\!"'\,
> > > \+ strerror  => 'errno => use locale; local $\! = $\_\[0\]; "$\!"'\,
> > >
> > > I attach two files&#8203;:
> > >
> > > one is a propose to test this bug\. it works for 5\.12\.2 and fails
> > > for
> > > 5\.21\.7
> > >
> > > another is a more complete test\, to show what I am doing\.
> >
> > A possible fix is attached\.
> 
> I think this will cause bug \#123171 to recur\.  locale\->import affects
> the currently\-compiling code\, so it needs to go in a BEGIN block&#8203;:
> BEGIN \{ local $\!; require locale; locale\->import\} local $\! = $\_\[0\];
> "$\!"\.  I haven’t tested any of this\.  Could you test your one\-liner
> from \#123171 and see whether strerror’s German output is affected by
> your patch?

Actually\, wrappers\.t in blead fails with a non\-English locale\, already before my patch\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/26/2014 08&#8203;:10 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Dec 26 06&#8203;:21&#8203;:08 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
>> Dana Pet 26\. Pro 2014\, 05&#8203;:29&#8203;:25\, tiago\.peczenyj@&#8203;gmail\.com reče&#8203;:
>>> Hello\,
>>>
>>> I find one bug on POSIX&#8203;::strerror on 5\.21\.7 \- it erase the $\!
>>> variable\!
>>>
>>> I discover this based on this ticket&#8203;:
>>>
>>> https://github.com/Weborama/Riak-Light/issues/18
>>>
>>> I try to reproduce the error using the 5\.21\.7 and I discover one
>>> strange behavior on POSIX&#8203;::strerror
>>>
>>> IF I do this
>>>
>>> \< perl code >
>>> use POSIX qw\(ETIMEDOUT strerror\);
>>> $\! = ETIMEDOUT;
>>> my $errmsg = strerror\(ETIMEDOUT\);
>>> \< / perl code >
>>>
>>> after the strerror\, the content of $\! was totally erased\. and it has a
>>> bad impact on my tests\.
>>>
>>> Seems the origin of this problem is this change between 5\.21\.6 and
>>> 5\.21\.7
>>>
>>> \- strerror  => 'errno => local $\! = $\_\[0\]; "$\!"'\,
>>> \+ strerror  => 'errno => use locale; local $\! = $\_\[0\]; "$\!"'\,
>>>
>>> I attach two files&#8203;:
>>>
>>> one is a propose to test this bug\. it works for 5\.12\.2 and fails for
>>> 5\.21\.7
>>>
>>> another is a more complete test\, to show what I am doing\.
>>
>> A possible fix is attached\.
>
> I think this will cause bug \#123171 to recur\.  locale\->import affects the currently\-compiling code\, so it needs to go in a BEGIN block&#8203;: BEGIN \{ local $\!; require locale; locale\->import\} local $\! = $\_\[0\]; "$\!"\.  I haven’t tested any of this\.  Could you test your one\-liner from \#123171 and see whether strerror’s German output is affected by your patch?
>

Your idea works for me\, as well as the attached\, more verbose patch\.  I 
like yours better\.

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0009-PATCH-perl-123503-Bug-in-POSIX-strerror-on-5.21.7.patch</summary>

```diff
From 966f6c0eaa9f5981484f3824c2438dd15e17bade Mon Sep 17 00:00:00 2001
From: Karl Williamson <khw@cpan.org>
Date: Fri, 26 Dec 2014 14:49:17 -0700
Subject: [PATCH 9/9] PATCH: [perl #123503] Bug in POSIX::strerror on 5.21.7

---
 ext/POSIX/lib/POSIX.pm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/POSIX/lib/POSIX.pm b/ext/POSIX/lib/POSIX.pm
index 5695990..9c7b33f 100644
--- a/ext/POSIX/lib/POSIX.pm
+++ b/ext/POSIX/lib/POSIX.pm
@@ -151,7 +151,7 @@ my %reimpl = (
     exit      => 'status => CORE::exit($_[0])',
     getenv    => 'name => $ENV{$_[0]}',
     system    => 'command => CORE::system($_[0])',
-    strerror  => 'errno => use locale; local $! = $_[0]; "$!"',
+    strerror  => 'errno => my $save_errno; BEGIN { $save_errno = $!; }; my $errstr; { use locale; $! = $_[0]; $errstr = "$!"; }; $!= $save_errno; $errstr',
     strstr    => 'big, little => CORE::index($_[0], $_[1])',
     chmod     => 'mode, filename => CORE::chmod($_[0], $_[1])',
     fstat     => 'fd => CORE::open my $dup, "<&", $_[0]; CORE::stat($dup)', # Gross.
-- 
1.9.1


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

To move the discussion forward\, I have applied Slaven's patch in a smoke branch&#8203;:

smoke\-me/jkeenan/slaven/123503\-POSIX

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @khwilliamson

This is now fixed by ec064ab75aa5b4dfd43af2f7654b9a0363ae02c4
All three ideas worked\, though the test was flawed because other things in the test file cause it to never fail\.  So I created a new test file for just the one test\, and verified that without the patch it failed\.

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
